engineering-plans/0002-subgraph-schema-merging.md

```diff
@@ -0,0 +1,214 @@
+- If the type is found, copy it and add a `@subgraphId("...")` directive.
+- If the type is imported with the `{ name: "", as: "" }` format, the merged type will include an `@originalName("...")` directive preserving the type name from the original schema.
+
+The `api_schema` function will add all the necessary types and fields for the imported types without requiring any changes.
+
+#### Example #1: Complete merge
+
+Local schema before calling `merged_schema`:
+
+```graphql
+type _Schema_
+  @import(
+    types: ["B"],
+    from: { id: "X" }
+  )
+
+type A @entity {
+  id: ID!
+  foo: B!
+}
+```
+
+Imported Schema X:
+
+```graphql
+type B @entity {
+  id: ID!
+  bar: String
+}
+```
+
+Schema after calling `merged_schema`:
+
+```graphql
+type A @entity {
+  id: ID!
+  foo: B!
+}
+
+type B @entity @subgraphId("X") {
+  id: ID!
+  bar: String
+}
+```
+
+#### Example #2: Incomplete merge
+
+Schema before calling `merged_schema`:
+
+```graphql
+type _Schema_
+  @import(
+    types: ["B"],
+    from: { id: "X" }
+  )
+
+type A @entity {
+  id: ID!
+  foo: B!
+}
+```
+
+Imported Schema X:
+
+```graphql
+NOT AVAILABLE
+```
+
+Schema after calling `merged_schema`
+
+```graphql
+type A @entity @subgraphId("...") {
+  id: ID!
+  foo: B!
+}
+
+type B @entity @placeholder {
+  id: ID!
+}
+```
+
+#### Example #3: Complete merge with `{ name: "...", as: "..." }`
+
+Schema before calling `merged_schema`
+
+```graphql
+type _Schema_
+  @imports(
+    types: [{ name: "B", as: "BB" }]
+    from: { id: "X" }
+  )
+
+type B @entity {
+  id: ID!
+  foo: BB!
+}
+```
+
+Imported Schema X:
+
+```graphql
+type B @entity {
+  id: ID!
+  bar: String
+}
+```
+
+Schema after calling `merged_schema`
+
+```graphql
+type B @entity {
+  id: ID!
+  foo: BB!
+}
+
+type BB @entity @subgraphId("X") @originalName("B") {
+  id: ID!
+  bar: String
+}
+```
+
+After the schema document is merged, the `api_schema` function will be called.
+
+### Cache Invalidation
+
+For each schema in the cache, keep a vector of subgraph pointers containing an element for each schema in the subgraph's import graph which was imported by name and the subgraph_id which was used during the schema merge. When a schema is accessed from the schema cache (and possibly only if this check hasn't happened in the last N seconds), check the current version for each of these schemas and run a diff against the versions used for the most recent schema merge. If there are any new versions, re merge the schema.
+
+Currently the `schema_cache` in the `Store` is a `Mutex<LruCache<SubgraphDeploymentId, SchemaPair>>`. A `SchemaPair` consists of two fields: `input_schema` and `api_schema`. To support the refresh flow, `SchemaPair` would be extended to be a `SchemaEntry`, with the fields `input_schema`, `api_schema`, `schemas_imported_by_name` (`Vec<(SubgraphDeploymentName, SubgraphDeploymentId)>`), and a `last_refresh_check` timestamp.
+
```

Comment:
I don't think `schemas_imported_by_name` is sufficient. Don't we also need to re-check imports from subgraph IDs that were not available in previous merges and later became available?

Perhaps it should be `schema_imports` or `schemas_imported` ?